### ISSUE TITLE
Compute nullspace of 2D dummies

### ISSUE THREAD

**Issue by lmmx** [lmmx/multicorr#2]
Folowing on from successfully implemented 2D dummies fb12651afb80e32f701ef7a59397b0d4581a3d67

```py
>>> data_with_dummies                                                          
   id c1 c2 c3     v  null_target_column  c1_B:c2_B:c3_C  c1_D:c2_D:c3_C
0   1  A  A  C  10.0                   0               0               0
1   2  B  A  C  20.0                   0               0               0
2   3  A  B  C  30.0                   0               0               0
3   4  B  B  C   NaN                   1               1               0
4   5  A  A  C  50.0                   0               0               0
5   6  B  A  C  60.0                   0               0               0
6   7  A  B  C  70.0                   0               0               0
7   8  B  B  C   NaN                   1               1               0
8   9  D  D  C   NaN                   1               0               1
```

The next step is to make the assertion pass again, which will require whittling down the extra unused degree of freedom (the extraneous `c3`)

_Originally posted by @lmmx in https://github.com/lmmx/multicorr/issues/1#issuecomment-1492979707_

**Comment by lmmx** [comment]
The non-empty powerset is denoted `P*(S)` and that's what we want to compute the columns (i.e. all combinations of columns) to find the most parsimonious multiple correlation (highest correlation with lowest column order)

**Comment by lmmx** [comment]
On reflection the powerset will be overkill: instead we only want the nullspace levels!

The nullspace will be all combinations of the levels in `data[null_indicator]`

The individual levels are given by

```py
>>> null_combos.apply(pd.Series.unique)
c1    [B, D]
c2    [B, D]
c3       [C]
dtype: object
```

So we can find the individual levels that are unique by simply taking the unique values of each column (factor), but to find the unique multi-level factors that are unique we are back to square one (suggesting we reuse the mechanism already made)

In fact we can begin with this [trivial] single level case by simply supplying just one factor at a time.

```py
>>> null_combos[cat_cols[:1]].astype(str).apply(dummy_2d, axis=1).fillna(0).astype(int)
   c1_B  c1_D
3     1     0
8     0     1
```

Or for all cats:

```py
>>> null_combos[[cat_cols[0]]].astype(str).apply(dummy_2d, axis=1).fillna(0).astype(int)
   c1_B  c1_D
3     1     0
8     0     1
>>> null_combos[[cat_cols[1]]].astype(str).apply(dummy_2d, axis=1).fillna(0).astype(int)
   c2_B  c2_D
3     1     0
8     0     1
>>> null_combos[[cat_cols[2]]].astype(str).apply(dummy_2d, axis=1).fillna(0).astype(int)
   c3_C
3     1
8     1
```

(Ignore the fact that this is not the most efficient way to do it)

So we have recapitulated the 5 levels of individual factors in the nullspace

We can make this generic as follows:

```py
from itertools import combinations

mono_combos = list(combinations(range(len(cat_cols)), r=1))
pd.concat([null_combos[[cat_cols[i] for i in combo]].astype(str).apply(dummy_2d, axis=1).fillna(0) for combo in mono_combos], axis=0).fillna(0).astype(int)
```
⇣
```py
   c1_B  c1_D  c2_B  c2_D  c3_C
3     1     0     0     0     0
8     0     1     0     0     0
3     0     0     1     0     0
8     0     0     0     1     0
3     0     0     0     0     1
8     0     0     0     0     1
```

The generic form of `cat_cols[i] for i in combo` is

```py
list(map(cat_cols.__getitem__, combo)) # or [*map(cat_cols.__getitem__, combo)]
```

**Comment by lmmx** [comment]
For all combinations now (not just `r=1` mono-level combinations):

```py
from itertools import chain, combinations

combos = [*chain.from_iterable(
  combinations(range(len(cat_cols)), r=r)
  for r in range(1, len(cat_cols)+1)
)]
nullspace_levelsets = [list(map(cat_cols.__getitem__, combo)) for combo in combos]
nullspace_combos = [
  null_combos[null_levels].astype(str).apply(dummy_2d, axis=1).fillna(0).astype(int)
  for null_levels in nullspace_levelsets
]
```

Confirmed gives correct results

```py
>>> pd.concat(nullspace_combos, axis=0).fillna(0).astype(int)
   c1_B  c1_D  c2_B  c2_D  c3_C  c1_B:c2_B  c1_D:c2_D  c1_B:c3_C  c1_D:c3_C  c2_B:c3_C  c2_D:c3_C  c1_B:c2_B:c3_C  c1_D:c2_D:c3_C
3     1     0     0     0     0          0          0          0          0          0          0               0               0
8     0     1     0     0     0          0          0          0          0          0          0               0               0
3     0     0     1     0     0          0          0          0          0          0          0               0               0
8     0     0     0     1     0          0          0          0          0          0          0               0               0
3     0     0     0     0     1          0          0          0          0          0          0               0               0
8     0     0     0     0     1          0          0          0          0          0          0               0               0
3     0     0     0     0     0          1          0          0          0          0          0               0               0
8     0     0     0     0     0          0          1          0          0          0          0               0               0
3     0     0     0     0     0          0          0          1          0          0          0               0               0
8     0     0     0     0     0          0          0          0          1          0          0               0               0
3     0     0     0     0     0          0          0          0          0          1          0               0               0
8     0     0     0     0     0          0          0          0          0          0          1               0               0
3     0     0     0     0     0          0          0          0          0          0          0               1               0
8     0     0     0     0     0          0          0          0          0          0          0               0               1
```

Note the redundant values:

```py
   c1_B  c1_D  c2_B  c2_D  c3_C  c1_B:c2_B  c1_D:c2_D  c1_B:c3_C  c1_D:c3_C  c2_B:c3_C  c2_D:c3_C  c1_B:c2_B:c3_C  c1_D:c2_D:c3_C
3     0     0     0     0     1          0          0          0          0          0          0               0               0
8     0     0     0     0     1          0          0          0          0          0          0               0               0
```

**Comment by lmmx** [comment]
The issue now is that we need to reverse the order of dummy col ("multicat name") checking

https://github.com/lmmx/multicorr/blob/73739f2d78747d63b141f1aa3c5795a24ed8e786/corr_analysis_multicat_dummy_form.py#L41-L46

Where we previously said

> "If the dummy columns have an exact match to the `name_multicat`-stringified version of the `cat_cols` subset of the row Series"

we now want to _reverse_ the comparison and say

> "If the `name_multicat`-stringified version of the `cat_cols` subset of the row Series has a partial (and not necessarily consecutive!) exact match to the dummy columns"

The non-consecutive part means that we cannot use substring `pd.Series.str.contains`

I expect what we have to do instead is compute the `name_multicat` for every combination.

Alternatively we can go in the opposite direction and compute the series subset to match against the row series (with the benefit that this would mean just 1 check operation per row)

In other words we are going to create a Series for each column (...i.e. we are creating a dataframe?) and will check if the row series is in the dataframe

A simpler way of looking at it is that since we know that every level in the nullspace is going to be in the dataset, then we're really checking for one or more columns in the data row being in the nullspace, and then simply combining those in all possible ways and creating the `name_multicat` for those combinations to get the dummy column name(s) to indicate at.

**Comment by lmmx** [comment]
I think the best approach is to

- compute a list of series (store them in a series if it's convenient, doesn't matter) from `unpack_multicat(dummy_col)` for all `dummy_cols` names
- retain this list while passing through all rows of the dataset
- compute equality comparison against each pre-computed series for the row (or rather the subset of columns of the row series to match those of the pre-computed series in each comparison). This might be facilitated by an `.all()` type comparison operator

Alternatively again, we could make single level dummies (potentially better vectorised) and use these in a comparison

**Comment by lmmx** [comment]
These approaches seem to be reimplementing the 2D check: we can first try doing the 2D check as-is by just using the `null_levelsets` in place of the `cat_cols` subsetting `row` in the `check_dummies_2d` function

**Comment by lmmx** [comment]
Successfully computed the nullspace powerset 2D dummies (or whatever we should call them!)

- 3241392

We still don't find the best result, and we get NaNs now in columns with no variance (which makes sense), namely the `c3_C` level we gave to all rows (deliberately).

```py
>>> full_dummies
   c1_B  c1_D  c2_B  c2_D  c3_C  c1_B:c2_B  c1_D:c2_D  c1_B:c3_C  c1_D:c3_C  c2_B:c3_C  c2_D:c3_C  c1_B:c2_B:c3_C  c1_D:c2_D:c3_C
0     0     0     0     0     1          0          0          0          0          0          0               0               0
1     1     0     0     0     1          0          0          1          0          0          0               0               0
2     0     0     1     0     1          0          0          0          0          1          0               0               0
3     1     0     1     0     1          1          0          1          0          1          0               1               0
4     0     0     0     0     1          0          0          0          0          0          0               0               0
5     1     0     0     0     1          0          0          1          0          0          0               0               0
6     0     0     1     0     1          0          0          0          0          1          0               0               0
7     1     0     1     0     1          1          0          1          0          1          0               1               0
8     0     1     0     1     1          0          1          0          1          0          1               0               1
>>> full_dummies.all()
c1_B              False
c1_D              False
c2_B              False
c2_D              False
c3_C               True
c1_B:c2_B         False
c1_D:c2_D         False
c1_B:c3_C         False
c1_D:c3_C         False
c2_B:c3_C         False
c2_D:c3_C         False
c1_B:c2_B:c3_C    False
c1_D:c2_D:c3_C    False
dtype: bool
```

Oh, actually in fact I modified the dataset, when the additional row is removed the assertion passes, feature complete :tada:
